Lection 13 HW

TheHsoa opened the pull request.

TheHsoa committed:
Lection 13 HW

Neliko commented on SerializeFractionCollection/SerializeFractionCollection/Program.cs:
```diff
@@ -1,15 +1,27 @@
+                fractionList.Add(new Fraction(random.Next(1, 10), random.Next(1, 100)));
+            }
+
+            fractionList.SerializeFractionCollection("outfile");
```
Круто, что сделан Extension. Не круто, что это сложно поддерживать) А что, если я попрошу тебя вывести не в файл, а в консоль? Или в БД?
Лучше заиспользовать отдельный класс, ответственный за сериализацию.
Тоже самое для десериализации.

P.S Я знаю, что в задании этого не требовалось. Но неплохо бы сразу писать легко изменяемый код)

Neliko replied:
При желании можно не делать :) но ты столкнёшься с этим же при изучении SOLID

TheHsoa replied:
Напишу еще одно расширение вывода в консоль или в БД
Конечно ты права, что делать расширения для такой узкой штуки как коллекция дробей не очень.
Подумаю что можно с этим сделать. Не очень понимаю что ты подразумеваешь под отдельным классом ответственным за сериализацию/десериализацию

TheHsoa replied:
По сути расширение и есть отдельный класс. Я могу сделать чтобы эти методы не были расширениями)
Можно прям 2 класса конечно сделать, тип в конструкторе на вход подаем путь к файлу
Кста я вроде не ловил ошибки, если файл не найден, занят и вообще когда все не так как хотелось бы

Neliko commented on SerializeFractionCollection/SerializeFractionCollection/Fraction.cs:
```diff
@@ -0,0 +1,57 @@
+
+        public override int GetHashCode()
+        {
+            unchecked
```
ооо, про такую штуку ничего не знаю)) было бы круто, если бы ты про неё мне рассказал :)

TheHsoa replied:
Понять бы вообще почему я именно так делаю...
Не смог пока найти вообще от куда взял этот GetHashCode()
Поход это все решарпер, подлюга.
http://stackoverflow.com/questions/102742/why-is-397-used-for-resharper-gethashcode-override

Neliko commented on SerializeFractionCollection/SerializeFractionCollection/Fraction.cs:
```diff
@@ -0,0 +1,57 @@
+
+        public Fraction(long numerator, long denominator)
+        {
+            if (denominator == 0) throw new DivideByZeroException("Denominator can't be zero");
```
Рррр. Ну ты понял,да?) И неплохо бы скобоньки добавить))

TheHsoa replied:
Добавлю, но эта строка довольно короткая и вообще я читал о то что в таких случаях допускается 3 варианта написания:
1. В одну строку
2. В две строки без скобок
3. В 4 строки со скобками

Я предположил что наиболее удачный будет вариант с одной строкой
Такой вариант не делает код не читаемым.

Neliko commented on SerializeFractionCollection/SerializeFractionCollection/Fraction.cs:
```diff
@@ -0,0 +1,57 @@
+            }
+
+            var runtimeType = GetType();
+            var left = obj.GetType();
```
Что такое left? вот тут я ничего не поняла

TheHsoa replied:
левая часть выражения
в общем то как сравнивать я украл с хабра
вор
https://habrahabr.ru/post/137680/

TheHsoa replied:
получаем типы сравниваемых объектов

TheHsoa replied:
в общем по той же статье с хабра чутка переделал всю эту вакханалию

Neliko commented on SerializeFractionCollection/SerializeFractionCollection/Fraction.cs:
```diff
@@ -0,0 +1,57 @@
+            return Equals(fraction);
+        }
+
+        protected bool Equals(Fraction other)
```
other название переменной мне совсем не нравится. Может otherFraction ?

TheHsoa replied:
переименовал

TheHsoa committed:
Правки по CR

Neliko commented on SerializeFractionCollection/SerializeFractionCollection/FractionCollectionSerializer.cs:
```diff
@@ -0,0 +1,32 @@
+
+        internal List<Fraction> DeserializeFractionCollection()
+        {
+            var binaryFormatter = new BinaryFormatter();
```
Слушай, а нам binaryFormatter точно нужно дважды инициализировать? Что изменится, если мы вынесем это дело в конструктор?

Neliko replied:
Слушай, а нам binaryFormatter точно нужно дважды инициализировать? Что изменится, если мы вытащим инициализацию? /

Neliko commented on SerializeFractionCollection/SerializeFractionCollection/FractionCollectionSerializer.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace SerializeFractionCollection
+{
+    internal class FractionCollectionSerializer
```
FractionCollectionSerializer - если сейчас хотим оставить такой класс, то следует его переименовать - всё же сейчас класс ограничен работой с файлом. В названии об этом нет ни слова.
А в целом можно пока остановиться и на такой реализации)

TheHsoa committed:
Правки по CR
Дополнительно поправил использование filestream по рекомендациям Microsoft и добавил исключения на работу с фалом
